docs/dev_cycle.part-04.rst

```diff
@@ -0,0 +1,315 @@
+
+.. code-block:: python
+
+    donations_usd = donator.call().donations_usd()
+    assert donations_usd == 1400
+
+
+OK, that wan't too complicated. Run the test:
+
+.. code-block:: shell
+
+    $ py.test --disable-pytest-warnings
+
+
+And the py.test results:
+
+.. code-block:: shell
+
+    platform linux -- Python 3.5.2, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
+    rootdir: /home/mary/projects/donations, inifile:
+    plugins: populus-1.8.0, hypothesis-3.14.0
+    collected 3 items
+
+    tests/test_donator.py ..F
+
+    ================================ FAILURES =======================================================
+    __________________________ test_usd_calculation _________________________________________________
+
+    chain = <populus.chain.tester.TesterChain object at 0x7f2736d1c630>
+
+        def test_usd_calculation(chain):
+
+            ONE_ETH_IN_WEI = 10**18  # 1 ETH == 1,000,000,000,000,000,000 Wei
+
+            donator, deploy_tx_hash = chain.provider.get_or_deploy_contract('Donator')
+            donator.transact({'value':ONE_ETH_IN_WEI}).donate(400)
+            donator.transact({'value':(2 * ONE_ETH_IN_WEI)}).donate(500)
+            donations_usd = donator.call().donations_usd()
+
+            # donated 1 ETH at $400 per ETH = $400
+            # donated 2 ETH at $500 per ETH = 2 * $500 = $1,000
+            # total $ value donated = $400 + $1,000 = $1,400
+    >       assert donations_usd == 1400
+    E       assert 1400000000000000000000 == 1400
+
+    tests/test_donator.py:32: AssertionError
+    ======================================= 1 failed, 2 passed, 15 warnings in 0.95 seconds =========
+
+
+Ooops. Something went wrong. But this is what tests are all about.
+
+Py.test tells us that the assert failed. Instead of 1,400, the ``donations_usd`` is 1400000000000000000000.
+And you know the saying: a billion here, a billion there, and pretty soon you're talking about real money.
+
+Where is the bug? you maybe guessed it already, but let's take a look at the contract's ``donate`` function:
+
+.. code-block:: solidity
+
+    function donate(uint usd_rate) public payable money_sent {
+        donations_total += msg.value;
+        donations_count += 1;
+        default_usd_rate = usd_rate;
+        uint in_usd = msg.value * usd_rate;
+        donations_usd += in_usd;
+        }
+
+Now it's clear:
+
+.. code-block:: solidity
+
+    uint in_usd = msg.value * usd_rate;
+
+This line multiplies ``msg.value``, which is in Wei, by ``usd_rate``, which is the exchange rate per *Ether*.
+
+Reminder: as of 0.4.17 Solidity does not have a workable decimal point calculation, and you have to handle fixed-point
+with integers. For the sake of simplicity, we will stay with ints.
+
+Edit the contract:
+
+.. code-block:: shell
+
+    $ nano contracts/Donator.sol
+
+
+And fix the line to:
+
+.. code-block:: solidity
+
+    uint in_usd = msg.value * usd_rate / 10**18;
```

Comment:
instead of `10**18` you can use a literal of `1 ether` and solidity will do the math for you.

Reply:
ok